src/units/abilities.cpp

```diff
@@ -1168,52 +1168,137 @@ void attack_type::modified_attacks(unsigned & min_attacks,
 	}
 }
 
-//Functions used for change damage_type list with damage
-static std::optional<std::string> select_damage_type(const unit_ability_list& abil_list, const std::string& type)
+//functions for select damage type who must be returned for each attribute used in damage_type()
+static std::string select_replacement_type(const unit_ability_list& damage_type_list)
 {
+	std::unordered_map<std::string, unsigned int> type_count;
+	unsigned int max = 0;
+	for(auto& i : damage_type_list) {
+		const config& c = *i.ability_cfg;
+		if(!c["replacement_type"].empty()) {
+			std::string type = c["replacement_type"].str();
+			unsigned int count = ++type_count[type];
+			if((count > max)) {
+				max = count;
+			}
+		}
+	}
+
+	if (type_count.empty()) return "";
+
 	std::vector<std::string> type_list;
-	for(auto& i : abil_list) {
-		if(!(*i.ability_cfg)[type].str().empty()){
-			type_list.push_back((*i.ability_cfg)[type].str());
+	for(auto& i : type_count){
+		if(i.second == max){
+			type_list.push_back(i.first);
 		}
 	}
-	if(type_list.size() >= 2){
+
+	if(type_list.size() > 1){
 		std::sort(type_list.begin(), type_list.end());
-		if(type_list.size() >= 3){
-			std::unordered_map<std::string, unsigned int> type_count;
-			for( const std::string& character : type_list ){
-				type_count[character]++;
+	}
+
+	if(!type_list.empty()){
+		return type_list.front();
+	}
+	return "";
+}
+
+static std::string select_alternative_type(const unit_ability_list& damage_type_list, unit_ability_list resistance_list, const unit& u)
+{
+	std::map<std::string, int> type_res;
+	//for don't have bug in showing in pre_attack window if opponent has resistance greater what 100% max_res shouldn't be initialised.
+	std::optional<int> max_res;
+	for(auto& i : damage_type_list) {
+		const config& c = *i.ability_cfg;
+		if(!c["alternative_type"].empty()) {
+			std::string type = c["alternative_type"].str();
+			if(!type_res[type]){
+				type_res[type] = u.resistance_value(resistance_list, type);
+				max_res = max_res ? std::max(*max_res, type_res[type]) : type_res[type];
+			}
```

Comment:
```suggestion
            int res = u.resistance_value(resistance_list, type);
			if(type_res.count(type)){
				type_res[type] = std::max(type_res[type], res);
			} else {
				type_res[type] = res;
			}
			max_res = max_res ? std::max(*max_res, res) : res;
```

Reply:
could you explain that?

Reply:
type_res.count(type) is equals to zero, right?

Reply:
My logic was what type_res[type] was edited what if don't already exist else it is not use to redone the calculation

Reply:
while what i don't know if ype_res.count(type) is equals to zero is eqhals or diffrent to 0, i can't commit these changes

Reply:
The logic should not be based on what value is in the map but whether there is a value or not. That is what `count(type)` checks. It's the same issue as your persistent empty checks. You keep checking the value of things when you really should be checking whether the thing actually exists.

Further the max should be calculated regardless if the type is in the map or not and the max for that type should be what ends up in the map.

Reply:
If the type is in map then value already added no. It is no use to check value in this case. Yoy suggest what resistance_value was called in each loop same if same type already checked

Reply:
How check if type is already in map or not

Reply:
That is what count does. I see you already adopted that part.

> If the type is in map then value already added no. It is no use to check value in this case. Yoy suggest what resistance_value was called in each loop same if same type already checked

You're saying the resistance value is always the same for the same damage type? I suppose nothing changes so that makes sense. Then I agree that what you have now is fine.

Reply:
If type is arcane and opponent has -10% resistance what arcane was cheked multiple time because two or more[damage_type]alternative_type=arcane opponent don'change resistance